### PR TITLE
Adding scalars to AxisManager

### PR DESCRIPTION
**Pull request by skhrg**
Adding scalar support to axis manager (#176 )

### CODE REVIEW

**Comment by kmharrington** [comment]
Thanks! 

I think @mhasself  is going to have a much better idea about how this implementation will work out than I will.

**Comment by mhasself** [comment]
Hi @skhrg  --- thanks for working on this.  I haven't tested this but it looks like it's on the right track.

Some notes on what else needs to happen here:
- I think the _assignments[] shouldn't be [None] for the scalar case, but rather [].  Scalars are "0-dimensional".
- Do something sensible for ``__repr__``.  For scalars we might just dump out var=<value of var>.
- Include scalar examples in the documentation.
- This needs to be included in the unit testing...

Here's the unit test I started (in test_core.py), for my attempt at this.
```
    def test_160_scalars(self):
        aman = core.AxisManager(core.LabelAxis('dets', ['a', 'b']),
                                core.OffsetAxis('samps', 100))
        aman.wrap_new('signal', shape=('dets', 'samps'))

        aman.wrap('x', np.array(12))
        aman.wrap('y', np.array(13))
        aman.wrap('z', np.array('hello'))
        aman.wrap('z2', np.array('hello'))

        # Accept trivially promoted scalars                                                             
        aman.wrap('x2', 12)
        aman.wrap('z3', 'hello')

        # Don't just let people wrap any old thing.                                                     
        with self.assertRaises(ValueError):
            aman.wrap('a_dict', {'a': 123})

```
It will a good idea to add some scalar entries in some of the other tests, e.g. test_140, and confirm it behaves as expected.

**Comment by skhrg** [comment]
@mhasself in your unit test was the intention that these lines:
```
        aman.wrap('x', np.array(12))
        aman.wrap('y', np.array(13))
        aman.wrap('z', np.array('hello'))
        aman.wrap('z2', np.array('hello'))
```
are wrapping scalars? In which case would you like me to treat an array with only a single element as a scalar?

**Comment by mhasself** [comment]
> @mhasself in your unit test was the intention that these lines:
> 
> ```
>         aman.wrap('x', np.array(12))
>         aman.wrap('y', np.array(13))
>         aman.wrap('z', np.array('hello'))
>         aman.wrap('z2', np.array('hello'))
> ```
> 
> are wrapping scalars?

Ahh... yes, those are probably not a good idea.  I was playing with the idea of wrapping scalars as 0-dimensional numpy arrays.  But that's perhaps confusing (np.isscalar(np.array(12)) is False).  You can drop those.

> In which case would you like me to treat an array with only a single element as a scalar?

Not totally sure what you mean here, but "no" is a safe answer!

People should be discouraged from using np.array(scalar) ... just use a scalar.

**Comment by skhrg** [comment]
> > In which case would you like me to treat an array with only a single element as a scalar?
> 
> Not totally sure what you mean here, but "no" is a safe answer!
> 
> People should be discouraged from using np.array(scalar) ... just use a scalar.

Yeah this answers my question, by array with a single element i just meant the length 1 array you would get from `np.array(scalar)`

File: docs/axisman.rst [hunk 1]
```diff
@@ -72,14 +72,22 @@ hold other AxisManagers, and some other stuff too).
 
     dset = core.AxisManager().wrap('tod', tod)
 
+AxisManagers can also hold scalars (ints, floats, strings, etc.), note that ``None`` is considered a scalar as well.
```
Comment (mhasself) [inline docs/axisman.rst:75]: For docs, please wrap lines at 78 chars or so (like the rest of the file).

What is included in "etc" in this line?  It's probably worth mentioning a scalar is anything that np.isscalar thinks is a scalar.

File: docs/axisman.rst [hunk 2]
```diff
@@ -72,14 +72,22 @@ hold other AxisManagers, and some other stuff too).
...
+
+.. code-block:: python
+
+    dset = dset.wrap('scalar', 1.0)
+
 Inspecting::
 
     >>> print(dset)
-    AxisManager(tod[3,10000])
+    AxisManager(tod[3,10000], scalar[0])
```
Comment (mhasself) [inline docs/axisman.rst:84]: The notation scalar[0] implies that scalar is 1-d array with shape (0,).  I think the repr in this case should just be
```
 AxisManager(tod[3,10000], scalar)
```

**Review by mhasself — CHANGES_REQUESTED** [review]
Sorry, just saw that this was marked ready.  Thanks again for doing this.

See a couple of comments inline about formatting.

Also... right now, this will fail:
```
 am.wrap('a', np.array([1]))
 am.wrap('a', np.array([2]))  # ValueError
```
but this succeeds:
```
 am.wrap('a', 1)
 am.wrap('a', 2)  # clobber
```
Although the errors are often annoying, the behavior should be consistent... so please get the scalar case to raise an error, like the array case.  Also please at tests for this behavior (in both scalar and array cases).

**Review by mhasself — APPROVED** [review]
Ok, thank you!

**Pull request merged**